### ISSUE TITLE
oshi vs sigar swap memory

### ISSUE THREAD

**Issue by battousai-dev** [oshi/oshi#1182]
im plan to migrate to oshi from hyperic sigar. i'm a little bit confused of the values of swap memory. here's the data i got from the different APIs:
hyperic sigar:
memory free=6533275648 
memory total=16157036544 
memory used=9623760896
swap free=6600974336
swap used=12508852224
swap total=19109826560

oshi: 
memory total: 16157036544
memory free: 6423027712
memory used: 9734008832
swap total: 2952790016
swap used: 84336640
swap free: 2868453376
memory page size in: 4096
memory swap page in:  5442129
memory swap page out: 75438

Sigar's swap total is the sum of oshi's memory total and oshi's swap total. (16157036544 + 2952790016 = ‭19,109,826,560‬). Is there a way to get the swap used using oshi's data?

**Comment by dbwiddis** [comment]
I guess it depends on your definition of "used".  As mentioned in #1175, OSHI calculates the percent of swap file "used" from the `“\Paging File\% Usage”` perfmon counter, described in the WMI documentation as the "Percentage of the page file instance in use. "   [Documentation for determining the page file size](https://docs.microsoft.com/en-us/windows/client-management/determine-appropriate-page-file-size) states:

> The \Paging File(*)% Usage performance counter measures the percentage of usage of each page file. 

I did go back to try to compare OSHI vs. the `systeminfo` command and it seems it does mismatch on the "used" vs. "in use" definition.

oshi:
"total":8511590400
"available":1412431872 --> used=7099158528 (?, see below)
"swapTotal":6442450944
"swapUsed":375062528
--> calculated total "Virtual memory" = 14954041344 = 14,261 MiB
--> calculated used "Virtual memory" = 7047593984 = **7,128 MiB**

This comes from: 
```
wmic path Win32_PerfRawData_PerfOS_PagingFile get Name,PercentUsage,PercentUsage_Base
Name                 PercentUsage  PercentUsage_Base
\??\C:\pagefile.sys  91568         1572864
```
Which calculates to 5% and matches the `perfmon` display for `% Usage`.

In contrast, `systeminfo` shows a higher number for `in use` but doesn't qualify what that really means.

systeminfo:
Total Physical Memory:     8,117 MB
Available Physical Memory: 1,357 MB
Virtual Memory: Max Size:  14,261 MB
Virtual Memory: Available: 5,528 MB
Virtual Memory: In Use:    **8,733 MB**

My Resource Monitor "in use" matches the "used" physical calculation above:
![image](https://user-images.githubusercontent.com/9291703/80519291-9b9ca980-893c-11ea-8658-db5c4c5b395c.png)

But it also shows 1134 MiB "Cached", an interesting feat since my swap file is only "using" 358 MiB.  I would guess that 1134 is just the physical portion of cached memory.  Adding it to the "in use" gives 7,899 MiB; adding that to my 358 MiB swap gives 8,257 MiB "Virtual Memory In Use", a lot closer to the `systeminfo` output.

A look at SIGAR's [source code](https://github.com/hyperic/sigar/blob/master/src/os/win32/win32_sigar.c) shows that it's retrieving the data from the [GlobalMemoryStatusEx](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-memorystatusex) function.  
```
        swap->total = memstat.ullTotalPageFile;
        swap->free  = memstat.ullAvailPageFile;
```
I considered using that, but the docs say: 
**ullTotalPageFile** - The current committed memory limit for the system or the current process, **whichever is smaller**, in bytes. To get the system-wide committed memory limit, call GetPerformanceInfo.

**ullAvailPageFile** - The maximum amount of memory the current process can commit, in bytes. This value is **equal to or smaller than the system-wide available commit value**. To calculate the system-wide available commit value, call GetPerformanceInfo and subtract the value of CommitTotal from the value of CommitLimit.

I just tested that and confirmed that this calculation in OSHI:
`pageSize * (performanceInfo.CommitLimit.longValue() - performanceInfo.CommitTotal.longValue())`
exactly matches the `systeminfo` output "Virtual Memory: Available", and would probably match SIGAR.

So, in summary:
 * I believe that OSHI's use of the `“\Paging File\% Usage”` perfmon counter is correct but differs from SIGAR and the `systeminfo` commands measuring total "Virtual Memory"
 * I believe SIGAR is incorrectly using `GlobalMemoryStatusEx` which gives a "current process" view of available memory rather than a system view.  This may not be significant.
 * The "Virtual Memory: Available" value from `systeminfo` can be obtained using `GetPerformanceInfo` function in JNA, using a modified version of the code in `WindowsGlobalMemory` to fetch the `CommitLimit` and `CommitTotal` fields, as shown above.  If it's important in your code to match SIGAR and `systeminfo` you can use JNA to grab these values.
 * I might consider adding "Virtual Memory: Max" and "Virtual Memory: Available" outputs to OSHI to match these outputs, in addition to the percent usage.

**Comment by dbwiddis** [comment]
So I'm adding the new fields as mentioned at the end of the above comment, and found this tidbit for "CommitTotal" (equal to Virtual Memory in use):

> Note that committing pages changes this value immediately; however, the physical memory is not charged until the pages are accessed.

This could explain the mismatch in the reports.  The memory has been committed by the system so it is charged against the "Virtual Memory in use".  But since it has not yet been accessed, it does not charge against the physical memory or swapfile.   It is "in use", but by whom?  It's neither physical nor swap, neither used nor not in use.  It's Schrödinger memory!
